Firestore/core/src/firebase/firestore/auth/user.h

```diff
@@ -69,6 +69,14 @@ inline bool operator!=(const User& lhs, const User& rhs) {
   return !(lhs == rhs);
 }
 
+// Specializations of std::hash is prohibited. We define a hash function to be
```

Comment:
Since we're using users as keys in unordered maps we need either a specialization of std::hash or a separate hash function like this one.

The advice against specializing std::hash is that we shouldn't compose them but this is just forwarding so it's safe.

We will eventually need to come up with a way to computing hashes for composite values but that day isn't today.

I vote we just specialize std::hash and make our usages of unordered_map easier.

Reply:
My concern is mostly with external users; I want them to not have to specify hash functions. For us / internal use, I don't care. I vote don't care. :)

(So if others are in favour, go ahead and specialize.)

Reply:
I am hesitating to code against style guide, which explicitly forbids specializing std::hash and provides a feasible workaround. It is possible later `User` gets changed and so does how we hash it (say someone decided to `is_authenticated` into the hash or add other class member variables). The style guide approach is more change-proof.